File: be/models/productSale.model.js
const db = require("../configs/db");

const productSaleSchema = new db.mongoose.Schema(
  {
    product_id: {
      type: db.mongoose.Schema.Types.ObjectId,
      ref: "productModel",
      required: true,
    },
    variant_id: {
      type: db.mongoose.Schema.Types.ObjectId,
      ref: "productVariantModel",
      default: null,
    },
    sale_name: { type: String, required: true },
    sale_price: { type: Number, required: true },
    sale_limit: { type: Number, default: null },
    sold_quantity: { type: Number, required: true, default: 0 },
    start_date: { type: Date, required: true },
    end_date: { type: Date, required: true },
    status: {
      type: String,
      required: true,
      enum: ["active", "inactive", "expired"],
      default: "active",
    },
    deleted_at: { type: Date, default: null },
  },
  { collection: "product_sales", timestamps: true },
);

let productSaleModel = db.mongoose.model("productSaleModel", productSaleSchema);
module.exports = { productSaleModel };

File: tuhubread/lib/models/product_sale.model.dart
import 'package:json_annotation/json_annotation.dart';

part 'product_sale.model.g.dart';

@JsonSerializable(fieldRename: FieldRename.snake)
class ProductSaleModel {
  @JsonKey(name: '_id')
  final String id;
  final String productId;
  final String? variantId;
  final String saleName;
  final double salePrice;
  final int? saleLimit;
  final int soldQuantity;
  final DateTime startDate;
  final DateTime endDate;
  final String status; // active / inactive / expired

  ProductSaleModel({
    required this.id,
    required this.productId,
    this.variantId,
    required this.saleName,
    required this.salePrice,
    this.saleLimit,
    required this.soldQuantity,
    required this.startDate,
    required this.endDate,
    required this.status,
  });

  /// Kiểm tra xem đợt sale có đang diễn ra không (countdown hợp lệ)
  bool get isActiveNow {
    final now = DateTime.now().toUtc();
    return status == 'active' &&
        now.isAfter(startDate.toUtc()) &&
        now.isBefore(endDate.toUtc()) &&
        (saleLimit == null || soldQuantity < saleLimit!);
  }

  factory ProductSaleModel.fromJson(Map<String, dynamic> json) =>
      _$ProductSaleModelFromJson(json);

  Map<String, dynamic> toJson() => _$ProductSaleModelToJson(this);
}
